### DISCUSSION TITLE
will `git add` support symlinks file

### DISCUSSION THREAD

**Discussion by hyaray** [git-for-windows/git#4737] [Q&A] [answered]
Hello everyone, 
there's a symlinks in dir, `git add .` will add a file with content path of source file, I want filetext,
is it possible?
my env: windows 10, cmder.

Best Regards!!

**Comment by dscho** [comment]
> there's a symlinks in dir, `git add .` will add a file with content path of source file, I want filetext,

I am afraid that I do not understand the question. Could you provide a step-by-step example?

**Reply by hyaray** [reply]
Assuming I have a file `d:\a\1.txt` with the content `hello`, and then I run git init in an empty folder `d:\b`. Following this, I create a symbolic link using `mklink d:\b\1.txt d:\a\1.txt`. Now, in the `b` folder, there is a `1.txt` file, which is a symbolic link.

When I run `git add . && git commit -am "update"` and later check the file content using `git show HEAD:1.txt`, I observe that the content is `d:\a\1.txt`—the path of the source file. However, my expectation is to see the content as the string `hello`, I hope this clarifies my scenario.

Feel free to use this description when responding on GitHub. If you have further details to add or if there's anything specific you'd like to emphasize, feel free to let me know.

**Reply by zrrbite** [reply]
Creating a hard-link with `mklink /H` might be what you're after.

Here I'm showing the contents of a file created with mklink (pointing to ../1/test.txt)
```
/d/dev/2 (master)
$ git show HEAD:test.txt
"hellotest"
```

After modifying the file i'ss symlinked to that is now registered as a change in /2 and i can print out the change:

```
/d/dev/2 (master)
$ git show HEAD:test.txt
changedokok ok "hello test"
```

But the problem is with someone checking out a symlink on their machine. e.g.

```
/d/dev  
$ git clone /d/dev/2/ 3.git                               
Cloning into '3.git'...                                   
done.    
```

Now even if i change test.txt in 1 it does not change in 3.git. I'd have to commit the changes in 2/ (because this is a real symlink on my machine) to get it in 3.git.

But what you could do, if you're after sharing of symlinks is to set up some sort of setup script that people can use, which creates the symlinks

**Reply by dscho** [reply, marked as answer]
> Assuming I have a file `d:\a\1.txt` with the content `hello`, and then I run git init in an empty folder `d:\b`. Following this, I create a symbolic link using `mklink d:\b\1.txt d:\a\1.txt`. Now, in the `b` folder, there is a `1.txt` file, which is a symbolic link.
> 
> When I run `git add . && git commit -am "update"` and later check the file content using `git show HEAD:1.txt`, I observe that the content is `d:\a\1.txt`—the path of the source file. However, my expectation is to see the content as the string `hello`,

`git show` only shows what Git stores, which is the target of the symbolic link. It does not resolve the symbolic link because that does not make sense, as there is no `D:\a\1.txt` _in the Git object database_ (which `git show` accesses).

To see the contents you expect to see, you need to check out the revision and then use a tool like `less` to look at the contents; This lookup will resolve the symbolic link on the filesystem, where the target path exists.
